### PR TITLE
Allow the FileTree to expand and collapse

### LINKED ISSUE #469: Enable the FileTree to fully expand and collapse
Depends on #185.

One of the keyboard shortcuts required for #199 is meant to expand the entire FileTree. I think it makes sense to allow the user to collapse the whole tree as well (possibly via the same keyboard shortcut, which would serve as a toggle). As far as I can tell, there's no simple way to just ask `react-treefold` to expand and/or collapse, the entire tree, so some code will need to be written to allow this to happen. Once that code is in place, we can have an action which can be dispatched to cause the tree to expand, and another to cause the tree to collapse.

I think it probably makes sense to complete #185 first, which I have raised for discussion in that issue.

### PR DESCRIPTION
**Pull request by bobsilverberg**
~~Depends on #503~~ 
Fixes #469

### CODE REVIEW

File: src/components/FileTree/index.tsx [hunk 1]
```diff
@@ -179,6 +202,12 @@ export class FileTreeBase extends React.Component<Props> {
           isNodeExpanded={this.isNodeExpanded}
           onToggleExpand={this.onToggleExpand}
         />
+        <button onClick={this.onExpandTree} type="button">
+          {gettext('Expand the tree')}
```
Comment (kumar303) [inline src/components/FileTree/index.tsx]: Something like *Open all folders* would probably make more sense in the UI

File: src/reducers/versions.spec.tsx [hunk 2]
```diff
@@ -178,6 +180,43 @@ describe(__filename, () => {
...
+    });
+
+    it('removes all paths from expandedPaths when collapseTree is dispatched', () => {
+      const version = fakeVersion;
+      let state = reducer(undefined, actions.loadVersionInfo({ version }));
+
+      const path = 'new/selected/path';
+      state = reducer(
+        state,
+        actions.toggleExpandedPath({ path, versionId: version.id }),
```
Comment (kumar303) [inline src/reducers/versions.spec.tsx]: I think this test should expand more than one path

File: src/reducers/versions.tsx [hunk 3]
```diff
@@ -624,6 +631,41 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
         },
       };
     }
+    case getType(actions.expandTree): {
+      const { versionId } = action.payload;
+
+      const version = state.versionInfo[versionId];
+      const { entries } = version;
+      const expandedPaths = entries
+        .filter((entry) => entry.type === 'directory')
```
Comment (kumar303) [inline src/reducers/versions.tsx]: I didn't see any tests that mix directories and files. That would be a good test.

File: src/components/FileTree/index.tsx [hunk 4]
```diff
@@ -179,9 +210,28 @@ export class FileTreeBase extends React.Component<Props> {
           isNodeExpanded={this.isNodeExpanded}
           onToggleExpand={this.onToggleExpand}
         />
+        <button id="openAll" onClick={this.onExpandTree} type="button">
```
Comment (kumar303) [inline src/components/FileTree/index.tsx]: Always use bootstrap components to keep the look and feel consistent.

I have a patch to suggest (below):
* Move the buttons up top where they are more effective. Without this, the *close all* button is inaccessible after opening everything.
* Render small, light buttons side by side to conserve space
* Truncate the button text on small screens. This just looks nicer but, also, *Open all ...* still makes enough sense.

```diff
diff --git a/src/components/FileTree/index.tsx b/src/components/FileTree/index.tsx
index 947dece..de12414 100644
--- a/src/components/FileTree/index.tsx
+++ b/src/components/FileTree/index.tsx
@@ -1,8 +1,9 @@
 import * as React from 'react';
-import { ListGroup } from 'react-bootstrap';
+import { Button, ListGroup } from 'react-bootstrap';
 import { connect } from 'react-redux';
 import Treefold, { TreefoldRenderProps } from 'react-treefold';
 
+import styles from './styles.module.scss';
 import FileTreeNode, {
   PublicProps as FileTreeNodeProps,
 } from '../FileTreeNode';
@@ -204,18 +205,34 @@ export class FileTreeBase extends React.Component<Props> {
 
     return (
       <ListGroup>
+        <div className={styles.controlButtons}>
+          <Button
+            className={styles.button}
+            id="openAll"
+            onClick={this.onExpandTree}
+            size="sm"
+            type="button"
+            variant="light"
+          >
+            {gettext('Open all folders')}
+          </Button>
+          <Button
+            className={styles.button}
+            id="closeAll"
+            onClick={this.onCollapseTree}
+            size="sm"
+            type="button"
+            variant="light"
+          >
+            {gettext('Close all folders')}
+          </Button>
+        </div>
         <Treefold
           nodes={[tree]}
           render={this.renderNode}
           isNodeExpanded={this.isNodeExpanded}
           onToggleExpand={this.onToggleExpand}
         />
-        <button id="openAll" onClick={this.onExpandTree} type="button">
-          {gettext('Open all folders')}
-        </button>
-        <button id="closeAll" onClick={this.onCollapseTree} type="button">
-          {gettext('Close all folders')}
-        </button>
       </ListGroup>
     );
   }
diff --git a/src/components/FileTree/styles.module.scss b/src/components/FileTree/styles.module.scss
new file mode 100644
index 0000000..01fb104
--- /dev/null
+++ b/src/components/FileTree/styles.module.scss
@@ -0,0 +1,14 @@
+.controlButtons {
+  display: grid;
+  grid-column-gap: 10px;
+  grid-template-columns: 1fr 1fr;
+  margin-bottom: 10px;
+  width: 100%;
+}
+
+.button {
+  overflow: hidden;
+  text-overflow: ellipsis;
+  white-space: nowrap;
+  width: 100%;
+}
```

Comment (bobsilverberg) [reply src/components/FileTree/index.tsx]: Thanks, that looks awesome!

**Comment by bobsilverberg** [comment]
Thanks @kumar303. I have addressed all of the comments. Please note that this is still built on top of #503.

File: src/reducers/versions.spec.tsx [hunk 5]
```diff
@@ -178,6 +180,67 @@ describe(__filename, () => {
...
+
+      expect(state).toHaveProperty(`versionInfo.${version.id}.expandedPaths`, [
+        path1,
+        path2,
+        getRootPath(createInternalVersion(version)),
+      ]);
+    });
+
+    it('does not add paths for files to expandedPaths when expandTree is dispatched', () => {
+      const path1 = 'path/1';
```
Comment (kumar303) [inline src/reducers/versions.spec.tsx]: This test would be easier to read if you used realistic examples, maybe like `scripts/` and `scripts/background.js`. Even though I understood the spec, I was having a hard time deciphering `path1` and `path2`.

**Pull request merged**